snap/snapcraft.yaml

```diff
@@ -68,72 +72,102 @@ parts:
         - libsodium23
         - libxml2
         - libxslt1.1
+        - libpython3.8-minimal
+        - libpython3.8-stdlib
+        - python3-distutils
+        - python3-pip
+        - python3-setuptools
+        - python3.8
         - squashfs-tools
         - xdelta3
+    build-environment:
+        - "CRYPTOGRAPHY_DONT_BUILD_RUST": "1"
     override-build: |
-      snapcraftctl build
+        snapcraftctl build
+
+        echo "Create libsodium symlink..."
+        TRIPLET_PATH="$SNAPCRAFT_PART_INSTALL/usr/lib/$(gcc -print-multiarch)"
+        LIBSODIUM="$(readlink -n "$TRIPLET_PATH/libsodium.so.23")"
+        ln -sf "$LIBSODIUM" "$TRIPLET_PATH/libsodium.so"
+
+        # Default to python3.8.
+        ln -sf python3.8 "$SNAPCRAFT_PART_INSTALL/usr/bin/python3"
 
-      echo "Create libsodium symlink..."
-      TRIPLET_PATH="$SNAPCRAFT_PART_INSTALL/usr/lib/$(gcc -print-multiarch)"
-      LIBSODIUM="$(readlink -n "$TRIPLET_PATH/libsodium.so.23")"
-      # Remove so the link can be recreated on re-builds
-      rm -f "$TRIPLET_PATH/libsodium.so"
-      ln -s "$LIBSODIUM" "$TRIPLET_PATH/libsodium.so"
+        # Site customization to load correct paths depending on running as snap or building.
+        cp "$SNAPCRAFT_PART_SRC/sitecustomize.py" "$SNAPCRAFT_PART_INSTALL/usr/lib/python3.8/sitecustomize.py"
+
+        # Install requisite python packages to dist-packages, which will be picked
+        # up by follow-on parts.  To install to lib/python3.8/site-packages, we
+        # need to set PYTHONUSERBASE and pass --user to install commands. --target
+        # could be used to install the package to the correct directory, but it would
+        # not available for pip operations.
+        export PYTHONUSERBASE="$SNAPCRAFT_PART_INSTALL"
+
+        # Pin pip to 20.1 to support broken python-apt on bionic.
+        $SNAPCRAFT_PART_INSTALL/usr/bin/python3 -m pip install -U pip==20.1 --user
+
+        # Install latest wheel & setuptools to support installing cryptography.
+        $SNAPCRAFT_PART_INSTALL/usr/bin/python3 -m pip install -U wheel setuptools --user
+
+        # Pin cryptography to 3.4 to avoid rust until it can be used on all arches.
+        $SNAPCRAFT_PART_INSTALL/usr/bin/python3 -m pip install -U cryptography==3.4 --user
```

Comment:
This is not really pinning, requirements will pick up a newer cryptography and then downgrade, essentially breaking builds. The new requirements.txt in master should make this line unnecessary though

Reply:
Dropped